File: engine/java/src/main/java/com/optionalpha/research/OptionsEquivalencyEngine.java
package com.optionalpha.research;

import java.util.HashMap;
import java.util.Map;

/**
 * Module AC2 (Java): Options Equivalency, Synthetics & Arbitrage Engine.
 * Implements Put-Call parity, basis arithmetic, and conversion/reversal arbitrage in Java.
 */
public class OptionsEquivalencyEngine {

    public Map<String, Double> computeBasis(double stockPrice, double interestRate, int daysToExp, double dividend) {
        double t = (double) daysToExp / 360.0;
        double carry = stockPrice * interestRate * t;
        double basis = carry - dividend;
        double fwd = stockPrice + basis;

        Map<String, Double> res = new HashMap<>();
        res.put("carry", Math.round(carry * 10000.0) / 10000.0);
        res.put("basis", Math.round(basis * 10000.0) / 10000.0);
        res.put("forward_price", Math.round(fwd * 10000.0) / 10000.0);
        return res;
    }

    public Map<String, Object> evaluatePutCallParity(
            double stockPrice, double strikePrice,
            double callPrice, double putPrice,
            double interestRate, int daysToExp, double dividend) {

        double t = (double) daysToExp / 360.0;
        double carry = stockPrice * interestRate * t;
        double basis = carry - dividend;

        double theoreticalStock = callPrice - putPrice + strikePrice - basis;
        double synthCall = stockPrice - strikePrice + putPrice + basis;
        double synthPut = callPrice + strikePrice - stockPrice - basis;

        Map<String, Object> out = new HashMap<>();
        out.put("actual_stock", stockPrice);
        out.put("theoretical_stock", Math.round(theoreticalStock * 100.0) / 100.0);
        out.put("synthetic_call", Math.round(synthCall * 100.0) / 100.0);
        out.put("synthetic_put", Math.round(synthPut * 100.0) / 100.0);
        return out;
    }

    public Map<String, Object> evaluateBoxSpread(double callSpreadCost, double putSpreadCost, double k1, double k2) {
        double boxCost = callSpreadCost + putSpreadCost;
        double parValue = Math.abs(k2 - k1);
        double profit = parValue - boxCost;

        Map<String, Object> res = new HashMap<>();
        res.put("box_cost", Math.round(boxCost * 100.0) / 100.0);
        res.put("par_value", Math.round(parValue * 100.0) / 100.0);
        res.put("guaranteed_profit", Math.round(profit * 100.0) / 100.0);
        res.put("is_arbitrage", profit > 0.05);
        return res;
    }
}

File: engine/julia/options_equivalency_engine.jl
# Module AC5 (Julia): Options Equivalency, Synthetics & Arbitrage Engine
# Scientific evaluation of Put-Call parity, forward carry, and box spread arbitrage

module OptionsEquivalencyEngine

export compute_basis, evaluate_parity, evaluate_box

function compute_basis(stock::Float64, rate::Float64, days::Int, div::Float64=0.0)
    t = Float64(days) / 360.0
    carry = stock * rate * t
    basis = carry - div
    fwd = stock + basis
    return (carry=round(carry, digits=4), basis=round(basis, digits=4), forward_price=round(fwd, digits=4))
end

function evaluate_parity(stock::Float64, strike::Float64, call::Float64, put::Float64, rate::Float64, days::Int, div::Float64=0.0)
    t = Float64(days) / 360.0
    basis = (stock * rate * t) - div
    theo_stock = call - put + strike - basis
    synth_call = stock - strike + put + basis
    synth_put = call + strike - stock - basis
    return (
        actual_stock = stock,
        theoretical_stock = round(theo_stock, digits=4),
        synthetic_call = round(synth_call, digits=4),
        synthetic_put = round(synth_put, digits=4)
    )
end

function evaluate_box(call_vert::Float64, put_vert::Float64, k1::Float64, k2::Float64)
    cost = call_vert + put_vert
    par = abs(k2 - k1)
    profit = par - cost
    return (
        box_cost = round(cost, digits=4),
        par_value = round(par, digits=4),
        profit = round(profit, digits=4),
        is_arbitrage = profit > 0.05
    )
end

end # module
